# Grocery Transactions — Data Exploration

Exploratory analysis of grocery transaction data using pandas.

**Dataset:** `grocery_database.xlsx` — transactions sheet  
**Goal:** Understand the structure, quality, and key patterns in the data.

## 1. Setup

In [5]:
# import libraries
import pandas as pd

## 2. Load Data

Loading the transactions sheet from the Excel source file.

In [13]:
# import data
df = pd.read_excel("../data/raw/grocery_database.xlsx", sheet_name='transactions')

## 3. Initial Inspection

Quick look at the shape, data types, and a sample of rows.

In [10]:
# check information about imported data
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 38506 entries, 0 to 38505
Data columns (total 6 columns):
 #   Column            Non-Null Count  Dtype         
---  ------            --------------  -----         
 0   customer_id       38506 non-null  int64         
 1   transaction_date  38506 non-null  datetime64[us]
 2   transaction_id    38506 non-null  int64         
 3   product_area_id   38506 non-null  int64         
 4   num_items         38506 non-null  int64         
 5   sales_cost        38506 non-null  float64       
dtypes: datetime64[us](1), float64(1), int64(4)
memory usage: 1.8 MB


In [12]:
# check shape
df.shape

(38506, 6)

In [6]:
# check top 5 data
df.head(5)

,customer_id,transaction_date,transaction_id,product_area_id,num_items,sales_cost
0,1,2020-04-10,435657533999,3,7,19.16
1,1,2020-04-10,435657533999,2,5,7.71
2,1,2020-06-02,436189770685,4,4,26.97
3,1,2020-06-02,436189770685,1,2,38.52
4,1,2020-06-10,436265380298,4,4,22.13


In [8]:
# check bottom 5 data
df.tail(5)

,customer_id,transaction_date,transaction_id,product_area_id,num_items,sales_cost
38501,870,2020-09-23,437316479610,5,4,51.61
38502,870,2020-09-23,437316479610,2,15,33.77
38503,870,2020-09-30,437385547544,5,3,23.46
38504,870,2020-09-30,437385547544,1,1,18.52
38505,870,2020-09-30,437385547544,2,6,2.63


In [9]:
# check random sample data
df.sample(25)

,customer_id,transaction_date,transaction_id,product_area_id,num_items,sales_cost
17853,392,2020-04-27,435825125179,3,35,103.88
7574,160,2020-04-04,435597871485,3,31,104.70
5208,109,2020-06-08,436247422124,3,35,90.50
1947,42,2020-04-10,435656282934,2,25,59.22
12295,266,2020-06-22,436389271700,4,3,12.51
3462,74,2020-04-24,435798207222,1,2,33.30
35464,806,2020-09-19,437278505637,3,2,7.57
13041,285,2020-07-04,436506342366,1,7,215.95
24272,540,2020-04-24,435795986491,5,3,28.40
32245,731,2020-05-25,436109291592,3,12,30.68


In [15]:
# check for 5% of the data
df.sample(frac=0.05)

,customer_id,transaction_date,transaction_id,product_area_id,num_items,sales_cost
27556,625,2020-09-18,437265054139,4,4,9.30
28110,637,2020-09-21,437292750332,3,6,24.12
25890,585,2020-04-16,435714675136,3,9,30.25
7379,157,2020-05-02,435876287475,3,6,14.33
37500,849,2020-09-23,437316892562,4,4,9.94
...,...,...,...,...,...,...
9033,197,2020-04-09,435641547635,4,1,6.06
27147,613,2020-07-15,436614020708,3,2,2.52
29063,660,2020-04-10,435654344632,5,5,41.74
12573,274,2020-07-29,436752803120,3,1,5.62


In [13]:
# check for missing data
df.isnull().sum()

customer_id         0
transaction_date    0
transaction_id      0
product_area_id     0
num_items           0
sales_cost          0
dtype: int64

In [16]:
df.describe()

,customer_id,transaction_date,transaction_id,product_area_id,num_items,sales_cost
count,38506.000000,38506,3.850600e+04,38506.000000,38506.000000,38506.000000
mean,429.884252,2020-06-30 19:37:03.798888,4.364726e+11,2.888303,6.190672,40.246398
min,1.000000,2020-04-01 00:00:00,4.355612e+11,1.000000,1.000000,0.000000
25%,209.000000,2020-05-16 00:00:00,4.360164e+11,2.000000,2.000000,11.420000
50%,422.000000,2020-07-01 00:00:00,4.364727e+11,3.000000,4.000000,23.180000
75%,656.000000,2020-08-16 00:00:00,4.369285e+11,4.000000,8.000000,45.980000
max,870.000000,2020-09-30 00:00:00,4.373898e+11,5.000000,310.000000,669.340000
std,255.218578,NaN,5.289576e+08,1.360284,5.961531,54.597118


In [19]:
df['product_area_id'].nunique()

5

In [22]:
# check for largest 25 data for sales_cost
df.nlargest(25,'sales_cost')

,customer_id,transaction_date,transaction_id,product_area_id,num_items,sales_cost
10443,224,2020-07-22,436683121376,1,27,669.34
1367,27,2020-08-31,437085904531,1,27,639.22
491,10,2020-09-29,437372244148,1,26,630.88
26319,593,2020-09-04,437126622541,1,24,611.83
32232,731,2020-04-29,435847289476,1,27,600.48
30124,682,2020-09-03,437112336056,1,29,589.26
19618,429,2020-07-27,436736081910,1,19,577.48
1610,33,2020-08-26,437037448340,1,25,568.09
10186,220,2020-08-22,436998525574,1,23,562.25
7081,150,2020-04-01,435562960558,2,310,557.73


In [24]:
# check for the largest 5 number of items
df.nlargest(5,'num_items')

,customer_id,transaction_date,transaction_id,product_area_id,num_items,sales_cost
7081,150,2020-04-01,435562960558,2,310,557.73
8829,193,2020-04-13,435684058573,3,79,219.59
10446,224,2020-07-25,436715473163,3,79,288.35
30079,682,2020-07-10,436563532172,3,62,207.95
15094,329,2020-06-30,436468977340,3,61,209.69


In [26]:
# check for the top 5 smallest sales_cost
df.nsmallest(5,'sales_cost')

,customer_id,transaction_date,transaction_id,product_area_id,num_items,sales_cost
14935,326,2020-09-07,437152966988,2,3,0.00
17926,392,2020-08-01,436789522435,2,6,0.00
19864,435,2020-04-26,435811839549,2,1,0.00
35057,796,2020-06-18,436342092911,2,7,0.00
10933,236,2020-05-06,435916318917,2,2,0.01


In [28]:
# check for the top 5 smallest number of items
df.nsmallest(5,'num_items')

,customer_id,transaction_date,transaction_id,product_area_id,num_items,sales_cost
51,2,2020-04-04,435593240741,5,1,12.42
52,2,2020-04-04,435593240741,1,1,19.79
58,2,2020-05-09,435946943100,4,1,7.07
59,2,2020-05-25,436106792154,5,1,5.08
80,2,2020-08-20,436972589589,2,1,0.21


### Observations

- **Shape:** 38,506 rows × 6 columns
- **Date range:** April 1 – September 30, 2020 (6 months of transactions)
- **Customers:** IDs range from 1–870, suggesting ~870 unique customers
- **Products:** 5 product areas (`product_area_id` 1–5 )
- **Items per transaction:** Median 4 items, max 310 — the max looks like an outlier worth investigating
- **Sales cost:** Median $23.18, mean $40.25 — mean is pulled up by high-value outliers (max $669.34). Min is $0.00 — worth checking if those are returns or data errors
- **Missing values:** None detected across all 6 columns (count = 38,506 for all)

**Next steps:**
- [ ] Confirm `transaction_id` is truly unique (no duplicate transactions)
- [ ] Investigate the `num_items` outlier (310) and `sales_cost` of $0.00
- [ ] Check `customer_id` — are all 870 IDs present or are there gaps?